1.1 


In [1]:
import numpy as np

house_A = np.array([85, 2, 5, 3])

house_B = np.array([120, 3, 10, 7])

house_C = np.array([200, 4, 2, 1])

print(house_A.shape)

(4,)


Câu B

Learning rate α = 0.01 là scalar vì:

chỉ có 1 giá trị duy nhất
không có nhiều chiều như vector

0.01 là scalar

Câu C 
Nếu thêm khoảng cách tới trung tâm thì vector có 5 đặc trưng 
vector thuộc R^5

1.2 

In [2]:
import numpy as np

u = np.array([1, 2, 3])
v = np.array([4, 5, 6])

uv_sum = u + v

print(uv_sum)

[5 7 9]


In [3]:
pred_model1 = np.array([0.2, 0.7, 0.5, 0.9])
pred_model2 = np.array([0.3, 0.6, 0.4, 0.8])

ensemble_avg = (pred_model1 + pred_model2) / 2

print(ensemble_avg)

[0.25 0.65 0.45 0.85]


Vị trí thứ 4 (0.85)

Tư duy

Ensemble thường tốt hơn vì:

giảm sai số của từng model
giảm overfitting
nhiều model cùng quyết định nên ổn định hơn

1.3 

In [4]:
import numpy as np

u = np.array([1, 2, 3])
c = 2

scaled = c * u

print(scaled)

[2 4 6]


In [5]:
W_old = np.array([0.5, -0.3, 0.8])

gradient = np.array([0.1, -0.2, 0.05])

alpha = 0.01

W_new = W_old - alpha * gradient

print(W_new)

[ 0.499  -0.298   0.7995]


Giải thích

Ta trừ gradient vì:

Gradient chỉ hướng loss tăng nhanh nhất.
Muốn giảm loss nên phải đi ngược hướng gradient.

1.4

In [6]:
import numpy as np

u = np.array([1, 2, 3])
v = np.array([4, 5, 6])

dot_uv = np.dot(u, v)

print(dot_uv)

32


TÍnh thủ công bằng cách: 1x4 + 2x5 + 3x6 =32

In [7]:
x = np.array([320, 8.5, 1.0])

w = np.array([0.002, 0.05, 0.03])

b = -0.5

y_pred = np.dot(w, x) + b

print(y_pred)

0.5950000000000002


1.5

In [10]:
import numpy as np

u = np.array([1, 2, 3, 4, 5])

l2_norm = np.linalg.norm(u)

l1_norm = np.linalg.norm(u, ord=1)

print(l2_norm)
print(l1_norm)

7.416198487095663
15.0


In [9]:
weights = np.array([0.8, -0.3, 0.05, 1.2, -0.9])

l2_penalty = 0.1 * np.linalg.norm(weights)**2

l1_penalty = 0.1 * np.linalg.norm(weights, ord=1)

print(l2_penalty)
print(l1_penalty)

0.29825
0.325


Giải thích

Lasso (L1):

phạt trực tiếp trị tuyệt đối của weights
dễ kéo weights nhỏ về đúng 0
giúp feature selection

Ridge (L2):

chỉ làm weights nhỏ lại
hiếm khi về đúng 0
giữ lại tất cả features

In [11]:
import numpy as np

a = np.array([1, 2, 3])
b = np.array([4, 5, 6])

cosine_sim = np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b))

print(cosine_sim)

0.9746318461970762


Giải thích:

Giá trị gần 1
Nghĩa là 2 vector rất giống nhau về hướng.

1.6

In [12]:
import numpy as np

X = np.array([
    [85, 8.2, 2],
    [92, 9.0, 3],
    [70, 7.5, 1],
    [88, 8.7, 4]
])

W = np.array([
    [0.4, 0.1],
    [0.3, 0.5],
    [0.2, 0.3]
])

Shape của $X \times W$ = (4, 2) > Giải thích: $(4, 3) \times (3, 2) \rightarrow (4, 2)$. Số cột của ma trận trước phải khớp với số hàng của ma trận sau.

Câu B — Tính toán Z và $X^T$1. Tính $Z = X @ W$:$$Z = \begin{bmatrix} 85 & 8.2 & 2 \\ 92 & 9.0 & 3 \\ 70 & 7.5 & 1 \\ 88 & 8.7 & 4 \end{bmatrix} \times \begin{bmatrix} 0.4 & 0.1 \\ 0.3 & 0.5 \\ 0.2 & 0.3 \end{bmatrix} = \begin{bmatrix} 36.86 & 13.2 \\ 40.1 & 14.6 \\ 30.45 & 11.05 \\ 38.61 & 14.35 \end{bmatrix}$$2. Tính $X^T$ (Chuyển vị):$$X^T = \begin{bmatrix} 85 & 92 & 70 & 88 \\ 8.2 & 9.0 & 7.5 & 8.7 \\ 2 & 3 & 1 & 4 \end{bmatrix}$$Shape thực tế của Z: (4, 2)Shape của $X^T$: (3, 4)

Trong Neural Network, $X @ W$ tương đương thao tác gì?Giải thích:Trong Neural Network, $X @ W$ tương đương với thao tác Forward Propagation (Lan truyền tiến) tại một layer.Ý nghĩa: Đây là bước tính tổng có trọng số (weighted sum) của các input.Thành phần: * $X$: Dữ liệu đầu vào (Input features).$W$: Trọng số (Weights) – thể hiện mức độ quan trọng của từng đặc trưng.Kết quả $Z$ chính là giá trị thô trước khi đưa qua hàm kích hoạt (activation function).

1.7

Câu A — Hadamard & Nhân ma trận1. Tích Hadamard (Nhân element-wise A * B):Nhân từng phần tử tương ứng của ma trận $A$ và $B$.$$hadamard = \begin{bmatrix} 1 \times 7 & 2 \times 8 & 3 \times 9 \\ 4 \times 1 & 5 \times 2 & 6 \times 3 \end{bmatrix} = \begin{bmatrix} 7 & 16 & 27 \\ 4 & 10 & 18 \end{bmatrix}$$2. Nhân ma trận (A @ B.T):Ma trận $A$ kích thước (2, 3) nhân với $B^T$ kích thước (3, 2) sẽ tạo ra ma trận (2, 2).$$mat\_mult = \begin{bmatrix} 1 & 2 & 3 \\ 4 & 5 & 6 \end{bmatrix} \times \begin{bmatrix} 7 & 1 \\ 8 & 2 \\ 9 & 3 \end{bmatrix} = \begin{bmatrix} 50 & 14 \\ 122 & 32 \end{bmatrix}$$

Câu B — Kiểm tra tính chất nghịch đảoKhi nhân một ma trận với ma trận nghịch đảo của chính nó ($M^{-1} \cdot M$), kết quả luôn là ma trận đơn vị $I$ (Identity Matrix).Kết quả M_inv @ M (sau khi làm tròn):$$M^{-1} M = \begin{bmatrix} 1.0 & 0.0 \\ 0.0 & 1.0 \end{bmatrix}$$

Câu C — Matrix Norms (L1 và L2/Frobenius)1. L1 Norm (Tổng giá trị tuyệt đối):$1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9 = \textbf{45}$2. L2 (Frobenius) Norm:Được tính bằng căn bậc hai của tổng bình phương tất cả các phần tử.$$L2 = \sqrt{1^2 + 2^2 + 3^2 + 4^2 + 5^2 + 6^2 + 7^2 + 8^2 + 9^2} = \sqrt{285}$$L2 (Frobenius) norm $\approx$ 16.88

In [13]:
import numpy as np

# ==========================================
# BÀI 1.6: Matrix Multiplication & Transpose
# ==========================================
print("--- BÀI 1.6 ---")
X = np.array([[85, 8.2, 2], [92, 9.0, 3], [70, 7.5, 1], [88, 8.7, 4]])
W = np.array([[0.4, 0.1], [0.3, 0.5], [0.2, 0.3]])

# Câu B: Tính Z và X.T
Z = X @ W
X_T = X.T

print(f"Z (X @ W):\n{Z}")
print(f"Shape của Z: {Z.shape}")
print(f"X Transpose:\n{X_T}")
print(f"Shape của X.T: {X_T.shape}\n")


# ==========================================
# BÀI 1.7: Hadamard, Inverse & Matrix Norms
# ==========================================
print("--- BÀI 1.7 ---")

# Câu A: Hadamard vs Matrix Mult
A = np.array([[1, 2, 3], [4, 5, 6]])
B = np.array([[7, 8, 9], [1, 2, 3]])

hadamard = A * B
mat_mult = A @ B.T

print(f"Hadamard (A * B):\n{hadamard}")
print(f"Matrix Multiplication (A @ B.T):\n{mat_mult}\n")

# Câu B: Inverse
M = np.array([[2.0, 1.0], [5.0, 3.0]])
M_inv = np.linalg.inv(M)
I = np.round(M_inv @ M, 8)

print(f"M_inv:\n{M_inv}")
print(f"M_inv @ M (Identity Matrix):\n{I}\n")

# Câu C: Matrix Norms
X_mat = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

l1_norm = np.sum(np.abs(X_mat))
l2_norm = np.linalg.norm(X_mat, ord='fro') # Hoặc np.sqrt(np.sum(X_mat**2))

print(f"L1 Norm: {l1_norm}")
print(f"L2 (Frobenius) Norm: {l2_norm:.2f}")

--- BÀI 1.6 ---
Z (X @ W):
[[36.86 13.2 ]
 [40.1  14.6 ]
 [30.45 11.05]
 [38.61 14.35]]
Shape của Z: (4, 2)
X Transpose:
[[85.  92.  70.  88. ]
 [ 8.2  9.   7.5  8.7]
 [ 2.   3.   1.   4. ]]
Shape của X.T: (3, 4)

--- BÀI 1.7 ---
Hadamard (A * B):
[[ 7 16 27]
 [ 4 10 18]]
Matrix Multiplication (A @ B.T):
[[ 50  14]
 [122  32]]

M_inv:
[[ 3. -1.]
 [-5.  2.]]
M_inv @ M (Identity Matrix):
[[1. 0.]
 [0. 1.]]

L1 Norm: 45
L2 (Frobenius) Norm: 16.88


2.1 

Hàm số f(x),Đạo hàm f′(x),Giá trị tại điểm cụ thể
f(x)=x2,f′(x)=2x,f′(3)=6
f(x)=3x+5,f′(x)=3,
f(x)=x3+2x2−x,f′(x)=3x2+4x−1,
f(x)=c,f′(x)=0,

In [14]:
import sympy as sp

# ==========================================
# CÂU A: Tính Đạo Hàm (Sử dụng Sympy)
# ==========================================
print("--- CÂU A: ĐẠO HÀM ---")
x = sp.Symbol('x')

# Danh sách các hàm số
functions = [
    x**2, 
    3*x + 5, 
    x**3 + 2*x**2 - x, 
    10  # hằng số c = 10
]

for func in functions:
    derivative = sp.diff(func, x)
    print(f"Hàm f(x) = {func}  =>  f'(x) = {derivative}")
    
    # Tính f'(3) riêng cho hàm đầu tiên
    if func == x**2:
        val = derivative.subs(x, 3)
        print(f"   - Tại x=3, f'(3) = {val}")

# ==========================================
# CÂU B: Gradient của MSE Loss
# ==========================================
print("\n--- CÂU B: GRADIENT CỦA MSE LOSS ---")

w = 0.5
x_val = 2.0
y_val = 1.0

# Bước 1: Tính y_hat (prediction)
y_hat = w * x_val

# Bước 2: Tính error
error = y_val - y_hat

# Bước 3: Tính dL/dw
# Công thức: dL/dw = -2 * x * (y - y_hat)
gradient = -2 * x_val * error

print(f"1. Prediction (y_hat) = {y_hat}")
print(f"2. Error (y - y_hat) = {error}")
print(f"3. dL/dw (Gradient)  = {gradient}")

# Giải thích hướng cập nhật w
if gradient > 0:
    print("\n=> Gradient dương: Cần GIẢM w để giảm Loss.")
elif gradient < 0:
    print("\n=> Gradient âm: Cần TĂNG w để giảm Loss.")
else:
    print("\n=> Gradient bằng 0: Đã đạt điểm cực tiểu (Loss tối ưu).")

--- CÂU A: ĐẠO HÀM ---
Hàm f(x) = x**2  =>  f'(x) = 2*x
   - Tại x=3, f'(3) = 6
Hàm f(x) = 3*x + 5  =>  f'(x) = 3
Hàm f(x) = x**3 + 2*x**2 - x  =>  f'(x) = 3*x**2 + 4*x - 1
Hàm f(x) = 10  =>  f'(x) = 0

--- CÂU B: GRADIENT CỦA MSE LOSS ---
1. Prediction (y_hat) = 1.0
2. Error (y - y_hat) = 0.0
3. dL/dw (Gradient)  = -0.0

=> Gradient bằng 0: Đã đạt điểm cực tiểu (Loss tối ưu).


 Giải thích về hướng của GradientNếu Gradient dương ($\frac{dL}{dw} > 0$): Ta nên GIẢM $w$ để giảm loss.Giải thích:Gradient cho biết hướng tăng nhanh nhất của hàm số. Nếu gradient dương, nghĩa là khi $w$ tăng thì Loss tăng.Để giảm Loss, ta phải đi ngược chiều gradient (quy tắc Gradient Descent): $w_{new} = w_{old} - \eta \cdot \frac{dL}{dw}$.Vì vậy, nếu $\frac{dL}{dw} > 0$, phép trừ sẽ làm giá trị $w$ giảm xuống.

In [15]:
import numpy as np

# --- Câu A: Partial Derivatives đơn giản ---
x_val, y_val = 2, 3
df_dx = 2 * x_val
df_dy = 2 * y_val

print(f"Câu A: df/dx = {df_dx}, df/dy = {df_dy}")

# --- Câu B: MSE với 2 tham số ---
y = 1.0
x1, x2 = 2.0, 3.0
w1, w2 = 0.3, 0.2

# 1. Tính prediction
y_hat = w1 * x1 + w2 * x2

# 2. Tính error
error = y - y_hat

# 3. Tính partial derivatives (Gradients)
dL_dw1 = -2 * x1 * error
dL_dw2 = -2 * x2 * error

print("\nCâu B:")
print(f"- Prediction (y_hat): {y_hat}")
print(f"- Error: {error}")
print(f"- Gradient w1 (dL/dw1): {dL_dw1:.2f}")
print(f"- Gradient w2 (dL/dw2): {dL_dw2:.2f}")

# Kiểm tra logic
if abs(dL_dw2) > abs(dL_dw1):
    print("\n=> dL/dw2 lớn hơn vì x2 > x1. Cần thay đổi w2 nhiều hơn w1 để giảm Loss.")

Câu A: df/dx = 4, df/dy = 6

Câu B:
- Prediction (y_hat): 1.2000000000000002
- Error: -0.20000000000000018
- Gradient w1 (dL/dw1): 0.80
- Gradient w2 (dL/dw2): 1.20

=> dL/dw2 lớn hơn vì x2 > x1. Cần thay đổi w2 nhiều hơn w1 để giảm Loss.


Giải thích:
Vì giá trị của đặc trưng $x_2$ ($3.0$) lớn hơn $x_1$ ($2.0$).
Trong công thức đạo hàm riêng của Loss function theo trọng số, ta có: $\frac{\partial L}{\partial w_i} = -2 \cdot \mathbf{x_i} \cdot error$. Khi $error$ là hằng số chung, trọng số ứng với input lớn hơn sẽ có gradient lớn hơn. Điều này có nghĩa là thay đổi nhỏ ở $w_2$ sẽ tác động đến tổng Loss mạnh mẽ hơn so với thay đổi ở $w_1$.

Câu A — Tính Partial Derivatives tại (x=2, y=3)Với hàm số $f(x,y) = x^2 + y^2$:$\frac{\partial f}{\partial x} = 2x$: Tại $x=2$, giá trị là $2 \times 2 = \mathbf{4}$$\frac{\partial f}{\partial y} = 2y$: Tại $y=3$, giá trị là $2 \times 3 = \mathbf{6}$Câu B — MSE với 2 tham sốCho dữ liệu: $y=1.0, x_1=2.0, x_2=3.0, w_1=0.3, w_2=0.2$Tính $\hat{y}$ (Prediction):$$\hat{y} = w_1x_1 + w_2x_2 = (0.3 \times 2.0) + (0.2 \times 3.0) = 0.6 + 0.6 = \mathbf{1.2}$$Tính error:$$error = y - \hat{y} = 1.0 - 1.2 = \mathbf{-0.2}$$Tính $\frac{\partial L}{\partial w_1}$:$$\frac{\partial L}{\partial w_1} = -2 \times x_1 \times error = -2 \times 2.0 \times (-0.2) = \mathbf{0.8}$$Tính $\frac{\partial L}{\partial w_2}$:$$\frac{\partial L}{\partial w_2} = -2 \times x_2 \times error = -2 \times 3.0 \times (-0.2) = \mathbf{1.2}$$Kết quả: $\frac{\partial L}{\partial w_1} = 0.8$ và $\frac{\partial L}{\partial w_2} = 1.2$.

Câu B — MSE với 2 tham sốCho dữ liệu: $y=1.0, x_1=2.0, x_2=3.0, w_1=0.3, w_2=0.2$Tính $\hat{y}$ (Prediction):$$\hat{y} = w_1x_1 + w_2x_2 = (0.3 \times 2.0) + (0.2 \times 3.0) = 0.6 + 0.6 = \mathbf{1.2}$$Tính error:$$error = y - \hat{y} = 1.0 - 1.2 = \mathbf{-0.2}$$Tính $\frac{\partial L}{\partial w_1}$:$$\frac{\partial L}{\partial w_1} = -2 \times x_1 \times error = -2 \times 2.0 \times (-0.2) = \mathbf{0.8}$$Tính $\frac{\partial L}{\partial w_2}$:$$\frac{\partial L}{\partial w_2} = -2 \times x_2 \times error = -2 \times 3.0 \times (-0.2) = \mathbf{1.2}$$Kết quả: $\frac{\partial L}{\partial w_1} = 0.8$ và $\frac{\partial L}{\partial w_2} = 1.2$.

2.3

In [16]:
import numpy as np

# Khởi tạo
x, y = 3.0, 4.0
alpha = 0.1
steps = 5

print(f"{'Bước':<6} | {'Điểm (x, y)':<20} | {'f(x,y)':<10} | {'Gradient':<15}")
print("-" * 60)

for i in range(steps + 1):
    f_val = x**2 + y**2
    grad = np.array([2*x, 2*y])
    
    print(f"{i:<6} | ({x:>.3f}, {y:>.3f}){' ':<7} | {f_val:>.3f}{' ':<5} | [{grad[0]:>.2f}, {grad[1]:>.2f}]")
    
    # Cập nhật Gradient Descent
    x = x - alpha * grad[0]
    y = y - alpha * grad[1]

Bước   | Điểm (x, y)          | f(x,y)     | Gradient       
------------------------------------------------------------
0      | (3.000, 4.000)        | 25.000      | [6.00, 8.00]
1      | (2.400, 3.200)        | 16.000      | [4.80, 6.40]
2      | (1.920, 2.560)        | 10.240      | [3.84, 5.12]
3      | (1.536, 2.048)        | 6.554      | [3.07, 4.10]
4      | (1.229, 1.638)        | 4.194      | [2.46, 3.28]
5      | (0.983, 1.311)        | 2.684      | [1.97, 2.62]


Nhận xét:Hướng hội tụ: Các điểm $(x, y)$ đang tiến dần về điểm $(0, 0)$.Tại sao đó là điểm tối ưu? * Tại $(0, 0)$, giá trị hàm số $f(0, 0) = 0$, đây là giá trị nhỏ nhất có thể của hàm tổng bình phương (Global Minimum).Tại đây, Gradient $\nabla f = [0, 0]$, nghĩa là thuật toán sẽ dừng lại vì không còn hướng nào để giảm thêm giá trị.

2.4

In [17]:
def relu(z):
    return max(0, z)

def relu_prime(z):
    return 1 if z > 0 else 0

# Dữ liệu đầu vào
x, w, b, y = 2.0, 0.5, 0.1, 1.0

# --- Forward Pass ---
z = w * x + b
a = relu(z)
loss = (y - a)**2

# --- Backward Pass (Chain Rule) ---
dl_da = -2 * (y - a)
da_dz = relu_prime(z)
dz_dw = x

dl_dw = dl_da * da_dz * dz_dw

print(f"--- Forward Pass ---")
print(f"z: {z:.2f}, a: {a:.2f}, Loss: {loss:.4f}")

print(f"\n--- Backward Pass ---")
print(f"dL/da: {dl_da:.2f}")
print(f"da/dz: {da_dz}")
print(f"dz/dw: {dz_dw}")
print(f"dL/dw (Gradient): {dl_dw:.2f}")

# Minh họa Dying ReLU
z_dead = -1.0
print(f"\nNếu z = {z_dead}: da/dz = {relu_prime(z_dead)} => dL/dw = 0 (Dying ReLU)")

--- Forward Pass ---
z: 1.10, a: 1.10, Loss: 0.0100

--- Backward Pass ---
dL/da: 0.20
da/dz: 1
dz/dw: 2.0
dL/dw (Gradient): 0.40

Nếu z = -1.0: da/dz = 0 => dL/dw = 0 (Dying ReLU)


Đây gọi là hiện tượng Dying ReLU. Khi một neuron rơi vào vùng giá trị âm, Gradient của nó triệt tiêu hoàn toàn. Do Gradient bằng 0, trọng số $w$ sẽ không bao giờ được cập nhật nữa (bị "chết"), khiến neuron đó ngừng học tập và không đóng góp gì cho mô hình.

3.1

In [18]:
import numpy as np

# Dữ liệu Câu B
scores = np.array([72, 85, 90, 68, 78, 95, 60, 82, 77, 88])

print("--- THỐNG KÊ LỚP CHÍNH ---")
print(f"Mean: {np.mean(scores)}")
print(f"Median: {np.median(scores)}")
print(f"Variance: {np.var(scores):.2f}")
print(f"Std Dev: {np.std(scores):.2f}")

# Dữ liệu Câu C
class_A = np.array([70, 75, 72, 78, 73])
class_B = np.array([50, 95, 60, 90, 55])

std_A = np.std(class_A)
std_B = np.std(class_B)

print("\n--- SO SÁNH HAI LỚP ---")
print(f"Std Dev Lớp A: {std_A:.2f}")
print(f"Std Dev Lớp B: {std_B:.2f}")

if std_B > std_A:
    print("=> Lớp B có Variance cao hơn, khó dự báo hơn cho ML.")

--- THỐNG KÊ LỚP CHÍNH ---
Mean: 79.5
Median: 80.0
Variance: 103.65
Std Dev: 10.18

--- SO SÁNH HAI LỚP ---
Std Dev Lớp A: 2.73
Std Dev Lớp B: 18.71
=> Lớp B có Variance cao hơn, khó dự báo hơn cho ML.


Câu C — So sánh 2 lớpTính toán nhanh độ lệch chuẩn ($\sigma$):Lớp A: Các điểm rất gần nhau $\rightarrow \sigma_A \approx \mathbf{2.8}$Lớp B: Các điểm cách xa nhau (phân cực) $\rightarrow \sigma_B \approx \mathbf{18.6}$💬 Trả lời câu hỏi:Lớp có Variance cao hơn: Lớp B.Ảnh hưởng đến ML: Lớp B "nguy hiểm" hơn cho model.Giải thích: Variance cao đồng nghĩa với dữ liệu có độ nhiễu (noise) lớn hoặc dải phân phối quá rộng. Model sẽ khó tìm ra quy luật chung (pattern) và dễ dẫn đến hiện tượng Overfitting (mô hình cố học các điểm cực đoan) hoặc sai số dự báo lớn. Dữ liệu đồng đều như Lớp A giúp model ổn định và hội tụ nhanh hơn.

3.2

In [19]:
import numpy as np

# Ma trận dữ liệu (Contingency Table)
# Hàng: Có KN (1) và Không KN (0)
# Cột: Giỏi Code (1) và Chưa giỏi (0)
data = np.array([[80, 30], 
                 [40, 50]])

total = data.sum()

# Xác suất cơ bản
p_code_1 = data[:, 0].sum() / total
p_exp_1 = data[0, :].sum() / total
p_joint = data[0, 0] / total

# Xác suất có điều kiện
p_code_given_exp = p_joint / p_exp_1
p_exp_given_code = p_joint / p_code_1

print(f"P(Code=1): {p_code_1:.2f}")
print(f"P(Exp=1): {p_exp_1:.2f}")
print(f"P(Code=1 | Exp=1): {p_code_given_exp:.3f}")
print(f"P(Exp=1 | Code=1): {p_exp_given_code:.3f}")

# Kiểm tra độc lập
print(f"\nP(A)*P(B) = {p_code_1 * p_exp_1:.2f} vs Thuc te = {p_joint:.2f}")

P(Code=1): 0.60
P(Exp=1): 0.55
P(Code=1 | Exp=1): 0.727
P(Exp=1 | Code=1): 0.667

P(A)*P(B) = 0.33 vs Thuc te = 0.40


Giỏi lập trình và kinh nghiệm có độc lập không?Không. Chúng có mối liên hệ mật thiết với nhau.Điều này có nghĩa là gì?Vì thực tế ($0.4$) lớn hơn kỳ vọng độc lập ($0.33$), ta thấy đây là một tương quan thuận. Nếu một sinh viên có kinh nghiệm, khả năng họ giỏi lập trình ($72.7\%$) cao hơn hẳn so với tỷ lệ giỏi lập trình trung bình của toàn bộ nhóm ($60\%$). Trong Machine Learning, biến Exp sẽ là một "predictor" (biến dự báo) rất tốt cho biến Code.

3.3

In [20]:
# Thông tin ban đầu
p_pos = 0.6
p_neg = 0.4
p_tuyet_given_pos = 0.7
p_tuyet_given_neg = 0.1

# Câu A: Law of Total Probability
p_tuyet = (p_tuyet_given_pos * p_pos) + (p_tuyet_given_neg * p_neg)

# Câu B: Bayes' Theorem
p_pos_given_tuyet = (p_tuyet_given_pos * p_pos) / p_tuyet
p_neg_given_tuyet = (p_tuyet_given_neg * p_neg) / p_tuyet

print(f"P('tuyệt'): {p_tuyet:.2f}")
print(f"P(Positive | 'tuyệt'): {p_pos_given_tuyet:.3f}")
print(f"P(Negative | 'tuyệt'): {p_neg_given_tuyet:.3f}")

if p_pos_given_tuyet > p_neg_given_tuyet:
    print("Kết quả phân loại: POSITIVE")
else:
    print("Kết quả phân loại: NEGATIVE")

P('tuyệt'): 0.46
P(Positive | 'tuyệt'): 0.913
P(Negative | 'tuyệt'): 0.087
Kết quả phân loại: POSITIVE


Ban đầu ($Prior$), khả năng một review là tích cực chỉ là 60%. Tuy nhiên, sau khi thấy từ "tuyệt" xuất hiện ($Likelihood$), niềm tin của mô hình vào việc đây là một review tích cực đã tăng vọt lên 91.3% ($Posterior$). Đây chính là cách Naive Bayes cập nhật kiến thức dựa trên bằng chứng (evidence).

3.4

In [21]:
import numpy as np
from scipy.stats import norm, binom

# --- Câu A: Bernoulli & Binomial ---
p = 0.3
n = 5
k = 2
prob_2_spam = binom.pmf(k, n, p)
print(f"P(2/5 email là spam): {prob_2_spam:.4f}")

# --- Câu B: Normal Distribution ---
mu = 70
sigma = 12

p_gt_80 = norm.sf(80, mu, sigma)  # sf = 1 - cdf
p_lt_50 = norm.cdf(50, mu, sigma)
top_10_threshold = norm.ppf(0.9, mu, sigma)

print(f"\nP(điểm > 80): {p_gt_80:.4f}")
print(f"P(điểm < 50): {p_lt_50:.4f}")
print(f"Ngưỡng top 10%: {top_10_threshold:.2f}")

# --- Câu C: Kiểm tra quy tắc 68-95-99.7 ---
# Tỷ lệ thực tế tính bằng code:
within_1std = norm.cdf(mu + sigma, mu, sigma) - norm.cdf(mu - sigma, mu, sigma)
within_2std = norm.cdf(mu + 2*sigma, mu, sigma) - norm.cdf(mu - 2*sigma, mu, sigma)

print(f"\nTỷ lệ thực tế 1 sigma: {within_1std:.2%}")
print(f"Tỷ lệ thực tế 2 sigma: {within_2std:.2%}")

P(2/5 email là spam): 0.3087

P(điểm > 80): 0.2023
P(điểm < 50): 0.0478
Ngưỡng top 10%: 85.38

Tỷ lệ thực tế 1 sigma: 68.27%
Tỷ lệ thực tế 2 sigma: 95.45%


3.5

In [22]:
import numpy as np

def calculate_stats(outcomes, probabilities):
    expected_value = np.sum(outcomes * probabilities)
    variance = np.sum(probabilities * (outcomes - expected_value)**2)
    return expected_value, variance

# Model A
acc_A = np.array([0.80, 0.82, 0.84, 0.83, 0.81])
prob_A = np.array([0.20, 0.30, 0.25, 0.15, 0.10])

# Model B
acc_B = np.array([0.70, 0.78, 0.88, 0.92])
prob_B = np.array([0.25, 0.35, 0.25, 0.15])

ea, vara = calculate_stats(acc_A, prob_A)
eb, varb = calculate_stats(acc_B, prob_B)

print(f"Model A: E = {ea:.4f}, Var = {vara:.6f}")
print(f"Model B: E = {eb:.4f}, Var = {varb:.6f}")

Model A: E = 0.8215, Var = 0.000203
Model B: E = 0.8060, Var = 0.006364


Câu C — Quyết định DeployModel được chọn: Model ALý do:Hiệu suất trung bình cao hơn: $E[A] (0.8215) > E[B] (0.806)$.Độ ổn định cao hơn (Rủi ro thấp hơn): $Var(A)$ nhỏ hơn rất nhiều so với $Var(B)$.Model A có các kết quả Accuracy rất tập trung quanh giá trị trung bình (từ $80\%$ đến $84\%$).Model B có biến động cực lớn (từ $70\%$ đến $92\%$). Dù đôi khi B đạt $92\%$, nhưng rủi ro rơi xuống $70\%$ là rất cao ($25\%$).

3.6

In [23]:
import numpy as np

gre = np.array([310, 325, 295, 335, 305])
admit = np.array([0.65, 0.80, 0.52, 0.90, 0.68])

# Câu A: Hiệp phương sai
# Lưu ý: ddof=0 để chia cho n thay vì n-1 (phù hợp tính thủ công ở trên)
cov_matrix = np.cov(gre, admit, ddof=0)
cov_val = cov_matrix[0, 1]

# Câu B: Hệ số tương quan
rho = np.corrcoef(gre, admit)[0, 1]

print(f"Mean GRE: {np.mean(gre)}")
print(f"Mean Admit: {np.mean(admit)}")
print(f"Covariance: {cov_val:.2f}")
print(f"Correlation (rho): {rho:.4f}")

Mean GRE: 314.0
Mean Admit: 0.7100000000000001
Covariance: 1.82
Correlation (rho): 0.9785


Ý nghĩa của $\rho$ và vai trò của FeatureGiải thích:$\rho > 0.5$ có nghĩa gì? Giá trị $\rho \approx 0.95$ cho thấy một tương quan thuận cực kỳ mạnh mẽ. Điều này có nghĩa là khi điểm GRE tăng, khả năng được nhận (Admit) cũng tăng theo một cách rất ổn định và tuyến tính.GRE có nên là feature quan trọng không? CÓ. Vì GRE có tương quan rất cao với mục tiêu dự báo (Admit), nó chứa rất nhiều "tín hiệu" giúp mô hình học được quy luật. Trong các mô hình Linear Regression, một feature có tương quan cao như thế này thường sẽ có trọng số ($w$) lớn và đóng vai trò chủ chốt trong kết quả dự báo.

Logistic regression 


Phần A — Lý thuyết & Công thức1. Sigmoid function$\sigma(z) = \frac{1}{1 + e^{-z}}$$\sigma(0) =$ $0.5$$\sigma(+\infty) \approx$ $1.0$$\sigma(-\infty) \approx$ $0$2. Binary Cross-Entropy LossKhi $\hat{y} = y = 1$: $L_i = -\log(1) =$ $0$ (Không có sai số).Khi $\hat{y} = 0.5, y = 1$: $L_i = -\log(0.5) \approx$ $0.693$ (Sai số lớn hơn).3. Gradient của BCE lossShape của $X^T @ error$ = (số_đặc_trưng, 1) — hoặc chính xác là bằng shape của vector trọng số $w$.

In [24]:
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

In [25]:
def bce_loss(y_true, y_pred):
    y_pred = np.clip(y_pred, 1e-10, 1 - 1e-10) # Tránh lỗi log(0)
    return -np.mean(y_true * np.log(y_pred) + (1 - y_true) * np.log(1 - y_pred))

In [26]:
# Forward pass:
z = X @ w + b                # Tính tổng có trọng số
y_hat = sigmoid(z)           # Chuyển thành xác suất qua Sigmoid
loss = bce_loss(y, y_hat)    # Tính sai số tổng hợp

# Backward pass (gradient):
error = y_hat - y            # Tính độ lệch giữa dự báo và thực tế
dw = (X.T @ error) / len(y)  # Đạo hàm theo trọng số w
db = np.mean(error)          # Đạo hàm theo bias b

# Update (Optimizer):
w = w - alpha * dw           # Cập nhật w theo hướng giảm loss
b = b - alpha * db           # Cập nhật b theo hướng giảm loss

ValueError: matmul: Input operand 1 does not have enough dimensions (has 0, gufunc core with signature (n?,k),(k,m?)->(n?,m?) requires 1)

Nhận xét (giả định theo mô hình logistic thông thường):Loss giảm từ: $\approx 0.7$ (ngẫu nhiên) xuống $< 0.1$ (tùy thuộc vào dữ liệu).Accuracy trên training set: $\approx$ $90\% - 100\%$ (nếu dữ liệu phân tách tốt). Nếu loss không giảm, nguyên nhân có thể là:Nguyên nhân 1: Learning rate ($\alpha$) quá lớn khiến trọng số bị nhảy vọt qua điểm tối ưu, hoặc quá nhỏ khiến mô hình học cực chậm.Nguyên nhân 2: Dữ liệu không được chuẩn hóa (Scaling). Nếu các feature có đơn vị quá lệch nhau (ví dụ 1 và 1.000.000), gradient sẽ bị mất ổn định.Nguyên nhân 3: Lỗi trong đạo hàm hoặc Forward pass. Ví dụ: Quên hàm Sigmoid khiến output không nằm trong khoảng $[0, 1]$, dẫn đến hàm Logarit bị lỗi.

Phần 1 — Linear AlgebraQ1. Dot product $w \cdot x$ có ý nghĩa gì?Ý nghĩa: Phép nhân vô hướng đại diện cho tổng có trọng số (weighted sum). Nó đo lường mức độ tương quan hoặc đóng góp của từng đặc trưng ($x$) vào kết quả dự báo thông qua các trọng số ($w$).Ví dụ: Dự báo giá nhà. $x = [\text{diện tích}, \text{số phòng}]$, $w = [50, 100]$. Giá nhà = $50 \times \text{diện tích} + 100 \times \text{số phòng}$.Q2. Tại sao Matrix Multiplication ($X@W$) hiệu quả hơn vòng for?Trả lời: Do kỹ thuật Vectorization (Vectơ hóa). Các thư viện như NumPy sử dụng tập lệnh SIMD của CPU và tối ưu hóa bộ nhớ đệm (cache), cho phép tính toán song song nhiều phần tử cùng lúc thay vì xử lý tuần tự từng dòng như vòng for.Q3. Transpose $X^T$ xuất hiện trong công thức nào?Công thức: Trong tính Gradient: $\frac{\partial L}{\partial w} = \frac{1}{n} X^T ( \hat{y} - y )$.Tại sao cần $T$: Để khớp kích thước ma trận (alignment). Nếu $X$ có shape $(n, m)$ và error có shape $(n, 1)$, ta cần $X^T$ shape $(m, n)$ để kết quả gradient có shape $(m, 1)$, khớp với hình dạng của vector trọng số $w$.Phần 2 — CalculusQ4. Gradient trỏ về hướng nào? Tại sao GD đi ngược lại?Hướng: Gradient luôn trỏ về hướng hàm số tăng nhanh nhất.Tại sao đi ngược: Vì mục tiêu của chúng ta là tối tiểu hóa (minimize) hàm Loss, nên ta phải đi về hướng hàm số giảm nhanh nhất.Q5. Chain Rule trong Backpropagation (3 hàm lồng nhau)Giải thích: Nếu $L = f(g(h(x)))$, đạo hàm của $L$ theo $x$ là tích của các đạo hàm từng chặng: $\frac{dL}{dx} = \frac{dL}{df} \cdot \frac{df}{dg} \cdot \frac{dg}{dh} \cdot \frac{dh}{dx}$.Trong Neural Network, thông tin lỗi được truyền ngược từ layer cuối về layer đầu bằng cách nhân các đạo hàm địa phương này lại với nhau.Q6. Vấn đề về Learning rate $\alpha$$\alpha$ quá lớn: Gây hiện tượng Overshooting (nhảy vọt qua điểm tối ưu), khiến hàm Loss dao động dữ dội hoặc không bao giờ hội tụ (diverge).$\alpha$ quá nhỏ: Mô hình học cực kỳ chậm, dễ bị kẹt ở các điểm cực tiểu địa phương (local minima) hoặc điểm yên ngựa (saddle points).Phần 3 — Probability & StatisticsQ7. Trong Bayes Theorem, $P(A)$ và $P(A|B)$ là gì?$P(A)$ - Prior (Xác suất tiên nghiệm): Niềm tin ban đầu của ta về sự kiện A khi chưa có bằng chứng B.$P(A|B)$ - Posterior (Xác suất hậu nghiệm): Niềm tin mới về A sau khi đã cập nhật bằng chứng B.Ý nghĩa: Cập nhật kiến thức cũ dựa trên thông tin mới thu thập được.Q8. Correlation cao có nghĩa là X gây ra Y không?Trả lời: Không. "Correlation is not causation" (Tương quan không phải là nhân quả).Giải thích: X và Y có thể cùng tăng/giảm do một biến ẩn thứ ba (Z) tác động, hoặc chỉ là sự trùng hợp ngẫu nhiên. Ví dụ: Doanh số bán kem và số vụ cháy rừng đều tăng vào mùa hè, nhưng kem không gây ra cháy rừng.Q9. Khi nào dùng $L_1$ (Lasso) thay vì $L_2$ (Ridge)?Trả lời: Khi bạn muốn lọc đặc trưng (feature selection). $L_1$ có xu hướng triệt tiêu hoàn toàn các trọng số không quan trọng về bằng 0, tạo ra mô hình thưa (sparse model). $L_2$ chỉ thu nhỏ trọng số chứ không triệt tiêu chúng.Q10. Tại sao Logistic Regression dùng Sigmoid thay vì output thẳng?Trả lời: Để thực hiện nhiệm vụ phân loại (classification).Hàm Sigmoid nén giá trị từ $(-\infty, +\infty)$ về đoạn $[0, 1]$, cho phép ta hiểu kết quả dưới dạng xác suất.Nó tạo ra một ngưỡng phân tách (decision boundary) phi tuyến tính giúp tách biệt các lớp dữ liệu hiệu quả hơn.